File: frontend/lib/api/knowledge.ts
import type { ApiEnvelope } from "../contracts";
import { authFetchJson } from "../api";
import type {
  KnowledgeAskResponse,
  KnowledgeEntitiesResponse,
  KnowledgeGraphResponse,
  KnowledgeTimelineResponse,
} from "../../types/knowledge";

export type KnowledgeEntitiesParams = {
  page?: number;
  page_size?: number;
  entity_type?: string;
  source_module?: string;
  q?: string;
};

export async function getKnowledgeEntities(
  clienteId: string,
  params: KnowledgeEntitiesParams = {},
): Promise<KnowledgeEntitiesResponse> {
  const query = new URLSearchParams();
  if (typeof params.page === "number" && params.page > 0) query.set("page", String(params.page));
  if (typeof params.page_size === "number" && params.page_size > 0) query.set("page_size", String(params.page_size));
  if (params.entity_type) query.set("entity_type", params.entity_type);
  if (params.source_module) query.set("source_module", params.source_module);
  if (params.q) query.set("q", params.q);

  const suffix = query.toString();
  const path = `/api/knowledge/${clienteId}/entities${suffix ? `?${suffix}` : ""}`;
  const response = await authFetchJson<ApiEnvelope<KnowledgeEntitiesResponse>>(path);
  return response.data;
}

export async function getKnowledgeGraph(clienteId: string): Promise<KnowledgeGraphResponse> {
  const response = await authFetchJson<ApiEnvelope<KnowledgeGraphResponse>>(`/api/knowledge/${clienteId}/graph`);
  return response.data;
}

export async function getKnowledgeTimeline(clienteId: string): Promise<KnowledgeTimelineResponse> {
  const response = await authFetchJson<ApiEnvelope<KnowledgeTimelineResponse>>(`/api/knowledge/${clienteId}/timeline`);
  return response.data;
}

export async function askKnowledge(
  clienteId: string,
  payload: { query: string; top_k?: number },
): Promise<KnowledgeAskResponse> {
  const response = await authFetchJson<ApiEnvelope<KnowledgeAskResponse>>(`/api/knowledge/${clienteId}/ask`, {
    method: "POST",
    body: JSON.stringify({
      query: payload.query,
      top_k: payload.top_k ?? 5,
    }),
  });
  return response.data;
}

File: backend/routes/knowledge.py
from __future__ import annotations

from typing import Any

from fastapi import APIRouter, Depends, Query

from backend.auth import authorize_cliente_access, get_current_user
from backend.schemas import (
    ApiResponse,
    KnowledgeAskRequest,
    KnowledgeEntityUpsertRequest,
    KnowledgeRelationUpsertRequest,
    UserContext,
)
from backend.services.knowledge_graph_service import build_graph
from backend.services.knowledge_service import (
    ask_with_event,
    ensure_knowledge_core_enabled,
    get_timeline,
    list_entities,
    upsert_entity,
    upsert_relation,
)
from backend.utils.database import get_session

router = APIRouter(prefix="/api/knowledge", tags=["knowledge"])


@router.post("/entities", response_model=ApiResponse)
def post_knowledge_entity(
    payload: KnowledgeEntityUpsertRequest,
    user: UserContext = Depends(get_current_user),
    session: Any = Depends(get_session),
) -> ApiResponse:
    ensure_knowledge_core_enabled()
    authorize_cliente_access(payload.cliente_id, user)
    saved = upsert_entity(
        session,
        payload.model_dump(),
        actor=user.display_name or user.sub,
    )
    return ApiResponse(data={"entity": saved})


@router.get("/{cliente_id}/entities", response_model=ApiResponse)
def get_knowledge_entities(
    cliente_id: str,
    entity_type: str | None = Query(None),
    source_module: str | None = Query(None),
    q: str | None = Query(None),
    page: int = Query(1, ge=1),
    page_size: int = Query(50, ge=1, le=200),
    user: UserContext = Depends(get_current_user),
    session: Any = Depends(get_session),
) -> ApiResponse:
    ensure_knowledge_core_enabled()
    authorize_cliente_access(cliente_id, user)
    data = list_entities(
        session,
        cliente_id=cliente_id,
        entity_type=entity_type,
        source_module=source_module,
        q=q,
        page=page,
        page_size=page_size,
    )
    return ApiResponse(data=data)


@router.post("/relations", response_model=ApiResponse)
def post_knowledge_relation(
    payload: KnowledgeRelationUpsertRequest,
    user: UserContext = Depends(get_current_user),
    session: Any = Depends(get_session),
) -> ApiResponse:
    ensure_knowledge_core_enabled()
    authorize_cliente_access(payload.cliente_id, user)
    saved = upsert_relation(
        session,
        payload.model_dump(),
        actor=user.display_name or user.sub,
    )
    return ApiResponse(data={"relation": saved})


@router.get("/{cliente_id}/graph", response_model=ApiResponse)
def get_knowledge_graph(
    cliente_id: str,
    max_entities: int = Query(500, ge=1, le=2000),
    max_relations: int = Query(1000, ge=1, le=5000),
    user: UserContext = Depends(get_current_user),
    session: Any = Depends(get_session),
) -> ApiResponse:
    ensure_knowledge_core_enabled()
    authorize_cliente_access(cliente_id, user)
    graph = build_graph(
        session,
        cliente_id=cliente_id,
        max_entities=max_entities,
        max_relations=max_relations,
    )
    return ApiResponse(data=graph)


@router.get("/{cliente_id}/timeline", response_model=ApiResponse)
def get_knowledge_timeline(
    cliente_id: str,
    limit: int = Query(100, ge=1, le=500),
    user: UserContext = Depends(get_current_user),
    session: Any = Depends(get_session),
) -> ApiResponse:
    ensure_knowledge_core_enabled()
    authorize_cliente_access(cliente_id, user)
    timeline = get_timeline(session, cliente_id=cliente_id, limit=limit)
    return ApiResponse(data=timeline)


@router.post("/{cliente_id}/ask", response_model=ApiResponse)
def post_knowledge_ask(
    cliente_id: str,
    payload: KnowledgeAskRequest,
    user: UserContext = Depends(get_current_user),
    session: Any = Depends(get_session),
) -> ApiResponse:
    ensure_knowledge_core_enabled()
    authorize_cliente_access(cliente_id, user)
    response = ask_with_event(
        session,
        cliente_id=cliente_id,
        query=payload.query,
        top_k=payload.top_k,
        actor=user.display_name or user.sub,
    )
    return ApiResponse(data=response)
